torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -924,8 +951,31 @@ void Reducer::finalize_backward() {
 
   // Wait for asynchronous reduction to complete and unflatten contents.
   for (auto& bucket : buckets_) {
-    TORCH_INTERNAL_ASSERT(bucket.work);
-    bucket.work->wait();
+    // See Note [DDP Communication Hook]
+    if (comm_hook_ == nullptr) {
+      TORCH_INTERNAL_ASSERT(
+          bucket.work,
+          "Expected bucket.work not to be null. "
+          "This may indicate that allreduce hooks were not properly installed.");
+      bucket.work->wait();
+    } else {
+      TORCH_INTERNAL_ASSERT(
+          bucket.future_work,
+          "Expected bucket.future_work not to be null. "
+          "This may indicate that communication hook was not properly installed.");
+      bucket.future_work->wait();
+
+      auto future_result =
+          comm_hook_->processFuture(bucket.future_work->value());
+
+      // Reinitialize bucket_views with the future_result by following
+      // the same logic in `inititalize_buckets`.
+      for (size_t i = 0; i < future_result.size(); i++) {
+        bucket.replicas[i].bucket_views.clear();
+
+        initialize_bucketviews(bucket.replicas[i], future_result[i]);
```

Comment:
I feel there is something wrong here. 

clear bucket_views does not free content storage, as replica.contents still point to the underneath storage. in this case, it could result in memory leak. 

future_result tensors are still pointing to the content storage right? so I assume the hook is operating bucket content tensors in place, and no need to copy back and forth outside of hook. Inside hook, you can do whatever you want though, e.g., copy bucket content tensors to a new tensor, then copy the new tensor back to bucket content tensor

Reply:
Any thoughts on this @pritamdamania ?

Reply:
@zhaojuanmao I'm not sure I completely followed the memory leak issue here. The original buckets stay alive until the Reducer is alive and once the reducer is destroyed the memory would be released (this is similar to what happens today). 

Now, if we create new tensors in our hook and have bucket_views point to that. In that case, when the next backward runs the memory for the previous Tensors would be released since now the bucket_views would point to a different set of Tensors.

Reply:
Here is what I understand: 

during construction time:   bucket tensors are allocated to hold whole grad copies. say 1GB. replica.bucket_views and replica.content (and grad in the near future) point to the bucket tensors. 

1st iteration: if the hook let replica.bucket_views point to new tensors, previous bucket tensors are still alive since replica.content is still point to them. **Then the memory usage is doubled, right?**  previous bucket tensors and new tensors are holding two copies of grads. 

Also in near future, we want to let grads point to bucket content tensor, expecting bucket tensors are mutated in place. if we go with this approach, we need to let grads explicitly point to new tensors in every iteration as well.